sympy/physics/mechanics/_geometry.py

```diff
@@ -0,0 +1,570 @@
+        surface in the ``N.y + N.z`` direction from ``pO``. These positions can
+        be set with:
+
+        >>> p1.set_pos(pO, r * N.x)
+        >>> p1.pos_from(pO)
+        r*N.x
+        >>> p2.set_pos(pO, r * (N.y + N.z).normalize())
+        >>> p2.pos_from(pO)
+        sqrt(2)*r/2*N.y + sqrt(2)*r/2*N.z
+
+        The geodesic length, which is in this case a quarter of the sphere's
+        circumference, can be calculated using the ``geodesic_length`` method:
+
+        >>> sphere.geodesic_length(p1, p2)
+        pi*r/2
+
+        If the ``geodesic_length`` method is passed an argument ``Point`` that
+        doesn't lie on the sphere's surface then a ``ValueError`` is raised
+        because it's not possible to calculate a value in this case.
+
+        Parameters
+        ==========
+
+        point_1 : Point
+            The point from which the geodesic length should be calculated.
+        point_2 : Point
+            The point to which the geodesic length should be calculated.
+
+        """
+        for point in (point_1, point_2):
+            if not self._point_is_on_surface(point):
+                msg = (
+                    f'Geodesic length cannot be calculated as point {point} '
+                    f'with radius {point.pos_from(self.point).magnitude()} '
+                    f'from the cylinder\'s center {self.point} does not lie on '
+                    f'the surface of {self} with radius {self.radius} and axis '
+                    f'{self.axis}.'
+                )
+                raise ValueError(msg)
+
+        relative_position = point_2.pos_from(point_1)
+        parallel_length = relative_position.dot(self.axis)
+
+        point_1_relative_position = point_1.pos_from(self.point)
+        point_1_perpendicular_vector = (
+            point_1_relative_position
+            - point_1_relative_position.dot(self.axis) * self.axis
+        ).normalize()
+
+        point_2_relative_position = point_2.pos_from(self.point)
+        point_2_perpendicular_vector = (
+            point_2_relative_position
+            - point_2_relative_position.dot(self.axis) * self.axis
+        ).normalize()
+
+        central_angle = _directional_atan(
+            point_1_perpendicular_vector
+                .cross(point_2_perpendicular_vector)
+                .dot(self.axis),
+            point_1_perpendicular_vector.dot(point_2_perpendicular_vector),
+        )
+
+        planar_arc_length = self.radius * central_angle
+        geodesic_length = _cancel_sqrt_of_squared_positive(
+            sqrt(parallel_length**2 + planar_arc_length**2)
+        )
+        return geodesic_length
+
+    def __repr__(self) -> str:
+        """Representation of a ``Cylinder``."""
+        return (
+            f'{self.__class__.__name__}(radius={self.radius}, '
+            f'point={self.point}, axis={self.axis})'
+        )
+
+
+def _cancel_sqrt_of_squared_positive(expr) -> Expr:
```

Comment:
Just a small detail I'd like to mention. The current implementation of this method will only work for symbols and not for `dynamicsymbols`:
```py
from sympy import posify
from sympy.physics.mechanics import dynamicsymbols
print(posify(sqrt(dynamicsymbols("a")**2)))
# prints (sqrt(a(_t)**2), {_t: t})
```

Reply:
I think that `posify` should basically never be used.

Reply:
I agree this is not ideal. Ideally I wanted to use assumptions to enforce the fact that radii must be positive. However, this is not possible to do and remain SymEngine compatible. This was the only approach I was able to implement which kept SymEngine compatibility.

Reply:
I think it is better to be very clear about what simplification you are trying to use and then make a function that does that. If the intention here is to simplify `sqrt(expr**2)` and you know that it will always be valid to do that then you can make a function that does that particular simplification and use that. Using `posify` will affect all symbols and including those that you did not expect and can cause many other simplifications that you might not want and that might not be valid.

More generally I think that it is better to avoid square roots altogether as much as possible. For example rather than saying
```python
is_inside = sqrt(x**2 + y**2) < r
```
it is better to say
```python
is_inside = x**2 + y**2 < r**2
```
The `sqrt` function as with all radicals is very problematic because of branch cuts. If there is any opportunity to avoid putting expressions involving symbols inside `sqrt` then it should always be taken.

Reply:
> cases while keeping SymEngine compatibility

This is a continual annoyance. I don't have any evidence that this is ever used by end users.